contracts/Staking.sol

```diff
@@ -0,0 +1,157 @@
+
+    /// @notice Store the users rewards per year
+    /// @dev Store the users rewards per year
+    mapping(address => uint256) public userRewardPerYear;
+
+    /// @notice Store the users staked amount tokens
+    /// @dev Store the users staked amount tokens
+    mapping(address => uint256) public staked;
+
+    /// @notice Initialize contract
+    /// @dev Initialize contract, sets reward/staking token addresses & staking info
+    /// @param rewardToken_ the reward token address
+    /// @param stakingToken_ the staking token address
+    constructor(address rewardToken_, address stakingToken_) {
+        require(
+            rewardToken_ != address(0) && stakingToken_ != address(0),
+            "ERROR_INVALID_ADDRESS"
+        );
+        rewardToken = IERC20(rewardToken_);
+        stakingToken = IERC20(stakingToken_);
+        stakingInfo.feePercentage = 40;
+        stakingInfo.cooldown = 10 days;
+        stakingInfo.stakingPeriod = 60 days;
+    }
+
+    /// @notice Calculate accrued reward tokens by staked tokens per year
+    /// @dev Calculate accrued reward tokens by staked tokens per year
+    /// @param amount_ the amount of staked tokens
+    /// @return the accrued reward tokens per year
+    function calculateRewardPerYear(uint256 amount_)
+        public
+        view
+        returns (uint256)
+    {
+        return (amount_ * (100 + apy)) / 100 - amount_;
+    }
+
+    /// @notice Add reward to contract for a specific period
+    /// @dev  Add reward to contract for a specific period
+    /// @param _start the start date of stake
+    /// @param _finish the finish date of stake
+    /// @param _rewardsAmount the amount of reward tokens
+    /// @param _apy the annual percentage yield
+    function setRewards(
+        uint256 _start,
+        uint256 _finish,
+        uint256 _rewardsAmount,
+        uint256 _apy
+    ) external virtual override onlyOwner {
+        require(
+            _start < _finish && block.timestamp < _start,
+            "Incorrect dates"
+        );
+        require(
+            _rewardsAmount != 0 && _rewardsAmount <= 500_000,
+            "Incorrect amount of tokens"
+        );
+        require(_apy != 0 && _apy <= 10, "Incorrect apy");
+
+        rewardToken.safeTransferFrom(msg.sender, address(this), _rewardsAmount);
+        startDate = _start;
+        endDate = _finish;
+        apy = _apy;
+        emit SetRewards(_start, _finish, _rewardsAmount, _apy);
+    }
+
+    /// @notice Transfer the amount of tokens from the user account and register staking for him
+    /// @dev Transfer the amount of tokens from the user account and register staking for him
+    /// @param _amount the amount of staked tokens
+    function stake(uint256 _amount) external virtual override {
+        require(block.timestamp < endDate, "The staking time is gone");
+        require(_amount > 0, "ERROR_AMOUNT_IS_ZERO");
+        require(totalSupply + _amount <= 5 * 10**6, "ERROR_MAX_TOTAL_SUPPLY");
```

Comment:
5*10**6 this is the Magic Number. Make a variable for more understanding

Reply:
totalSupply in 18 decimals by default then you compare with 6 decimals, it's not correct